pontoon/pretranslation/utils.py

```diff
@@ -0,0 +1,122 @@
+
+    locale_entities = list(locale_entities)
+
+    translations = []
+
+    locale_list = []
+    translatedresource_dict = {}
+    projectlocale_list = []
+
+    for locale in locales:
+        for entity in entities:
+            locale_entity = '{}-{}'.format(locale.id, entity.id)
+            if not locale_entity in locale_entities:
+                string = ""
+                tm_response = translation_memory_util(
+                    text=entity.string,
+                    locale=locale,
+                    max_results=5,
+                )
+
+                if tm_response and int(float(tm_response[0]['quality'])) == 100:
+                    string = tm_response[0]['target']
+
+                elif locale.google_translate_code:
+                    gt_response = google_translate_util(
+                        text=entity.string,
+                        locale_code=locale.google_translate_code,
+                    )
+
+                    if 'translation' in gt_response.keys():
+                        string = gt_response['translation']
+
+                if string:
+                    t = base.models.Translation(
+                        entity=entity,
+                        locale=locale,
+                        string=string,
+                        user=user,
+                        date=now,
+                        approved=False,
+                        active=True,
+                    )
+                    translations.append(t)
+                    locale_resource = '{}-{}'.format(locale.id, entity.resource.id)
+
+                    if not locale_resource in translatedresource_dict.keys():
+                        translatedresource = base.models.TranslatedResource.objects.get(
+                            resource=entity.resource,
+                            locale=locale
+                        )
+                        translatedresource_dict[locale_resource] = translatedresource
+
+
+
+
+    translations = base.models.Translation.objects.bulk_create(translations)
+
+    # update latest activity once and for all
+    if len(translations)>0:
+        project.latest_translation = translations[0]
+        project.save(update_fields=['latest_translation'])
+
+    # for locale in locale_list:
+    #     locale.save(update_fields=['latest_translation'])
+
+    # for projectlocale in projectlocale_list:
+    #     projectlocale.save(update_fields=['latest_translation'])
+
+    # bulk_update(locale_list, update_fields=['latest_translation'])
+    # bulk_update(projectlocale_list, update_fields=['latest_translation'])
+    # bulk_update(translatedresource_dict.values(), update_fields=['latest_translation'])
+
+    for translation in translations:
+        translation.update_latest_translation()
+
+    for translatedresource in translatedresource_dict.values():
+        # translatedresource.save(update_fields=['latest_translation'])
+        translatedresource.calculate_stats()
```

Comment:
I don't have measurements/metrics, but every `calculate_stats` call takes some time (I think you should measure how long it takes to recalculate all stats for your test project/locale).
I don't want to overshoot with the optimization (@mathjazz?). If recalculation of the stats will be the slowest part of pre-translation, you can take a look at https://bugzilla.mozilla.org/show_bug.cgi?id=1344658.
The general idea for the optimization in this case would be to generate the difference between stats and later apply them to translated resources, instead of full recalculation.

Reply:
OTOH, [sync](https://github.com/mozilla/pontoon/blob/2346e09c916b42c628bed59f4a8fd6e03a33a18d/pontoon/sync/core.py#L219) just calls calculate_stats on each changed resource, too.

We could however use something like [batch_actions](https://github.com/mozilla/pontoon/blob/2346e09c916b42c628bed59f4a8fd6e03a33a18d/pontoon/batch/views.py#L39) do.

Reply:
This was already on my bucket list for the weekend and I was going to discuss this in #pontoon Thanks a lot for the heads up @jotes.